lib/_stream_readable.js

```diff
@@ -425,9 +406,11 @@ Readable.prototype.read = function(n) {
   const state = this._readableState;
   const nOrig = n;
 
-  // If we're asking for more than the current hwm, then raise the hwm.
+  // If we're asking for more than the current hwm, then raise the hwm to
+  // the next highest power of 2 to prevent increasing hwm excessively in
+  // tiny amounts.
   if (n > state.highWaterMark)
-    state.highWaterMark = computeNewHighWaterMark(n);
+    state.highWaterMark = Math.pow(2, Math.ceil(Math.log(n) / Math.log(2)));
```

Comment:
Why did you remove the previous method? I would rather keep the previous logic. Is the new version faster?

(We might need to add some checks anyway to ceil it to the max value of an Integer, as we do not want to turn this into a double).

Reply:
Because it was magic and I don't think it needs to be fast here.

> We might need to add some checks anyway to ceil it

Then we do need a MAX_HWM but it probably be something closer to MAX_SAFE_INTEGER and if larger than that we should error?

Reply:
> Because it was magic and I don't think it needs to be fast here.

Byte operations are not magic. I would retain them because several people spent time in optimizing those, and I bet this hit some hot path that none of us can see right now.

> Then we do need a MAX_HWM but it should probably be something closer to MAX_SAFE_INTEGER and if larger than that we should error?

Agreed. I would not throw in this PR so we can back port it down to 10 if we think it’s a good idea. A follow up PR that enforces it is ok.

Reply:
Maybe something lower than that, e.g. 500MB or a GB. The reason is that I’m not sure Node can read that amount of data as a single chunk with an encoding set (it requires a 2x on allocated memory).